static/src/javascripts/projects/common/utils/steady-page.js

```diff
@@ -0,0 +1,187 @@
+                    // this prevents excessive and jarring scrolling
+                    return go(assign(state, {
+                        prevHeight: state.prevHeight + state.newHeight
+                    }));
+                }
+            });
+
+        return promise;
+
+    }
+
+    /**
+     * Insert an element into the page
+     * Use if your element doesn't exist and is inserted into a container
+     * ** Don't use fastdom - it is handled in this utility **
+     * @param {HTMLElement} container The element that the component is being inserted into
+     * @param {Function} cb Should contain all functionality that displays and lays-out the element
+     */
+    function insert(container, cb) {
+        if (!config.switches.steadyPageUtil) {
+            return fastdom.write(cb);
+        }
+
+        var initialState = {
+            scrollY: window.scrollY,
+            prevHeight: 0
+        };
+
+        q.enqueue({
+            container: container,
+            cb: cb
+        });
+
+        return (running ? promise : go(initialState)).then(scrollThePage);
+    }
+
+
+    /**
+     * Given a batch, call all of the callbacks on the insertion object
+     *
+     * @param  {Array} batch
+     */
+    function insertElements (batch) {
+        return fastdom.write(function () {
+            batch.forEach(function (insertion) {
+                insertion.cb();
+            });
+        });
+    }
+
+    function scrollThePage (scrollY) {
+        if (scrollY) {
+            window.scrollTo(0, scrollY);
+            running = false;
+        }
+    }
+
+    /**
+     * Given a batch and a previous currentBatchHeight, measure the height of each container
+     * in the batch
+     *
+     * @param  {Array} batch
+     */
+    function getHeightOfAllContainers (batch) {
+        return fastdom.read(function() {
+            // Add all the heights of the passed in batch
+            // removing the current height
+            return batch.filter(elementIsAbove).reduce(function(height, insertion) {
+                return height + readHeight(insertion.container);
+            }, 0);
+        });
+
+        function elementIsAbove(el) {
+            var parentElArray = [];
+            var parentEl = el.container;
+            var viewportAdjustment = Math.max(document.documentElement.clientHeight, window.innerHeight || 0) / 2;
+
+            // Push the container's parents into an array so we can calculate
+            // the position of the elements all the way until the body
+            while (parentEl) {
+                parentElArray.push(parentEl);
+                parentEl = parentEl.offsetParent;
+            }
+
+            // Get the top position of the element + half of the viewport (to account for an element loading
+            // just below the top of the viewport)
+            var elTopPos = parentElArray.reduce(function(topPos, parentEl){
```

Comment:
I just love complexity 💥 Seriously though, for some reason I thought I needed to get the element from the top of the page, but yes, I only care about the viewport. Doh. Updated.